### PR TITLE
Implement Python xDS interop client.

### PR DESCRIPTION
**Pull request by gnossen**
I discovered a couple of existing shortcomings while implementing this
client that mean this is *not yet ready for inclusion in CI*. I
nonetheless want to get an early review and integrate this to the master
branch, in the interest of small easily reviewable code changes.

The first problem was that the bazel build has apparently never pulled
roots.pem in as a data dependency. This appears not to have been a
problem until XDS enters into the mix. This has been added into the
Bazel build with a small change to the pyx_library rule.

The larger problem is that there is currently no public Python API to
get a peer's identity from the client side. This is crucial to
determining the correctness of an interaction of a client-server pair
under XDS. I intend to add such an method to the RpcContext interface in
an upcoming PR and use it in the interop client. For the moment, I
simply fake the peer details.

Finally, I add a knob to run_xds_tests.py allowing multiple instances of
this test to exist at once. Multiple instances forwarding to the same
port cannot exist on GCE, so this enables multiple people to run tests
at the same time.

### CODE REVIEW

File: src/python/grpcio/grpc/_cython/BUILD.bazel [hunk 1]
```diff
@@ -2,13 +2,21 @@ package(default_visibility = ["//visibility:public"])
...
+)
+
 pyx_library(
     name = "cygrpc",
     srcs = glob([
         "**/*.pxi",
         "cygrpc.pxd",
         "cygrpc.pyx",
     ]),
+    data = [":copy_roots_pem"],
```
Comment (lidizheng) [inline src/python/grpcio/grpc/_cython/BUILD.bazel]: nit: We have credentials in https://github.com/grpc/grpc/tree/master/src/python/grpcio_tests/tests/unit/credentials. We could reuse them. It might be complex, and it is okay to be complete in separate PR.

File: src/python/grpcio_tests/tests/interop/xds_interop_client.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+            end = start + duration_per_query
+            call, _ = stub.UnaryCall.with_call(messages_pb2.SimpleRequest(),
+                                               timeout=float(
+                                                   args.rpc_timeout_sec))
+            print(f"Got result {request_id}")
+            sys.stdout.flush()
+            with _global_lock:
+                for watcher in _watchers:
+                    # TODO: Implement a peer details getter.
+                    peer = f"192.168.1.{request_id % 255}"
```
Comment (ericgribkoff) [inline src/python/grpcio_tests/tests/interop/xds_interop_client.py]: The peer for the `LoadBalancerStatsResponse.rpcs_by_peer` should actually come from the server's response (so `SimpleResponse.hostname`, as done [here](https://github.com/grpc/grpc/blob/5312a3aeb2c06e6d8c518365e4a701456bd653c3/test/cpp/interop/xds_interop_client.cc#L142)). Getting the peer from the context (ala `context.peer()` in the C++ client) is only used if `-print_response` is true, which is only used by an internal test that should soon be deprecated, so I don't think there's any need for Python to implement this flag or pull the peer from the context at all.

File: src/python/grpcio_tests/tests/interop/xds_interop_client.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            if args.print_response:
+                if call.code() == grpc.StatusCode.OK:
+                    print("Successful response.")
+                    sys.stdout.flush()
+                else:
+                    print(f"RPC failed: {call}")
+                    sys.stdout.flush()
+            now = time.time()
+            while now < end:
+                time.sleep(end - now)
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/interop/xds_interop_client.py]: nit: Token bucket?

File: src/python/grpcio_tests/tests/interop/xds_interop_client.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+                                ):
+
+    def __init__(self):
+        super(_LoadBalancerStatsServicer).__init__()
+
+    def GetClientStats(self, request: messages_pb2.LoadBalancerStatsRequest,
+                       context: grpc.ServicerContext
+                      ) -> messages_pb2.LoadBalancerStatsResponse:
+        print("Received stats request.")
+        sys.stdout.flush()
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/interop/xds_interop_client.py]: nit: Should we use `logging`?

File: src/python/grpcio_tests/tests/interop/xds_interop_client.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+# TODO: Back with a LoadBalancerStatsResponse proto?
+class _StatsWatcher:
+    _start: int
+    _end: int
+    _rpcs_needed: int
+    _rpcs_by_peer: DefaultDict[str, int]
+    _no_remote_peer: int
+    _lock: threading.Lock
+    _condition: threading.Condition
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/interop/xds_interop_client.py]: nit: `threading.Condition` creates a lock if got given. We could use that one `with self._condition:`.

File: src/python/grpcio_tests/tests/interop/xds_interop_client.py (hunk 3 above)
Comment (gnossen) [reply src/python/grpcio_tests/tests/interop/xds_interop_client.py]: Having a bursty traffic pattern would differ from the other clients and require a more complicated implementation. Unless there's a compelling reason for it, I'd prefer not to.

Comment (lidizheng) [reply src/python/grpcio_tests/tests/interop/xds_interop_client.py]: Feel free to implement either way. I don't have a strong preference. A plain implementation can be:

```python
token = 1.0
last = time.monotonic()

while ...:
    if token < 1:
        time.sleep(1/args.qps * (1-token))
    now = time.monotonic()
    token = min(2, token - 1 + args.qps * (now-last))
    last = now
    ...
```

Comment (gnossen) [reply src/python/grpcio_tests/tests/interop/xds_interop_client.py]: Discussed with Eric. We need a steady QPS. That is, as close as possible to one RPC every `1 / QPS` seconds. That rules out a token bucket approach. Still, this is pretty cool. I wasn't aware of this algorithm. I'll have to keep this one in my back pocket.

File: src/python/grpcio/grpc/_cython/BUILD.bazel (hunk 1 above)
Comment (gnossen) [reply src/python/grpcio/grpc/_cython/BUILD.bazel]: Those are test credentials. Note that the directory says that they're not meant for production. Not that this change is to the production `cygrpc` target. We needed this in order to support creds callbacks in Bazel since the beginning. This flaw just wasn't noticed until xDS started requiring it.

Sorry for resurrecting this thread 24 days later. I just now got back onto this PR.

File: src/python/grpcio_tests/tests/interop/xds_interop_client.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        futures: Dict[int, grpc.Future] = {}
+        while not _stop_event.is_set():
+            request_id = None
+            with _global_lock:
+                request_id = _global_rpc_id
+                _global_rpc_id += 1
+            start = time.time()
+            end = start + duration_per_query
+            _start_rpc(request_id, stub, float(args.rpc_timeout_sec), futures)
+            _remove_completed_rpcs(futures, args.print_response)
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/interop/xds_interop_client.py]: optional: Sweep and remove works. An alternative can be a factory method and `add_done_callback`, which calls callback automatically and we can focus on the scheduling part.

File: src/python/grpcio_tests/tests/interop/xds_interop_client.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
+def _run(args: argparse.Namespace) -> None:
+    logger.info("Starting python xDS Interop Client.")
+    global _global_server  # pylint: disable=global-statement
+    channel_threads: List[threading.Thread] = []
+    for i in range(args.num_channels):
+        thread = threading.Thread(target=_run_single_channel, args=(args,))
+        thread.start()
+        channel_threads.append(thread)
+    _global_server = grpc.server(futures.ThreadPoolExecutor())
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/interop/xds_interop_client.py]: nit: Are we support Python 2.7?

Comment (gnossen) [reply src/python/grpcio_tests/tests/interop/xds_interop_client.py]: Nope. No need. This is a binary.

File: tools/internal_ci/linux/grpc_xds_bazel_test_in_docker.sh [hunk 8]
```diff
@@ -45,12 +45,24 @@ touch "$TOOLS_DIR"/src/proto/grpc/testing/__init__.py
     "$PROTO_SOURCE_DIR"/messages.proto \
     "$PROTO_SOURCE_DIR"/empty.proto
 
-bazel build test/cpp/interop:xds_interop_client
+export GRPC_VERBOSITY=debug
+export GRPC_TRACE=xds_client,xds_resolver,cds_lb,xds_lb
 
-GRPC_VERBOSITY=debug GRPC_TRACE=xds_client,xds_resolver,cds_lb,xds_lb "$PYTHON" \
```
Comment (ericgribkoff) [inline tools/internal_ci/linux/grpc_xds_bazel_test_in_docker.sh]: This should still have verbose logging enabled.

More importantly, I'm not sure about combining Python and C++ in the same job. The runs for this test already take long enough that I don't think we want the same kokoro job to execute C++ and Python sequentially. Python should probably just have its own stand-alone xDS Kokoro job and corresponding script here on Github - at the least, C++ and Python should only be in the same job if they are configured to separately report status for Python and C++ for dashboards etc.

File: src/python/grpcio_tests/tests/interop/xds_interop_client.py (hunk 6 above)
Comment (gnossen) [reply src/python/grpcio_tests/tests/interop/xds_interop_client.py]: I gave this a shot, but found that the callback isn't happening in the case of timeouts. Looks like a bug. Considering the difference is cosmetic, I'm going to stick with the sweep approach for this PR and fix [the bug](https://github.com/grpc/grpc/issues/22387) in a separate PR.

File: tools/internal_ci/linux/grpc_xds_bazel_test_in_docker.sh (hunk 8 above)
Comment (gnossen) [reply tools/internal_ci/linux/grpc_xds_bazel_test_in_docker.sh]: I think I've split things up the way you want (along with a pending internal change). PTALA.

Comment (gnossen) [reply tools/internal_ci/linux/grpc_xds_bazel_test_in_docker.sh]: C++ tests unrenamed as desired. PTALA.

**Comment by gnossen** [comment]
Failures:
 - https://github.com/grpc/grpc/issues/22409
 - https://github.com/grpc/grpc/issues/22341
